### ISSUE TITLE
[tracking issue] streams

### ISSUE THREAD

**Issue by yoshuawuyts** [async-rs/async-std#129]
With #125 out, it's probably worth looking at which other parts of [`std::iter`](https://doc.rust-lang.org/std/iter/index.html) we can port to [`async_std::stream`](https://docs.rs/async-std/0.99.4/async_std/stream/index.html). This issue is intended to track what's left for us to port.

## Missing free functions
- [ ] `from_fn`
- [ ] `repeat_with`
- [ ] `successors`
- [ ] `once_with`

## Missing traits
- [ ]  `DoubleEndedStream`
- [ ] `ExactSizeStream`
- [ ] `Extend`
- [ ] `FusedStream`
- [ ] `Product`
- [ ] `Sum`

## Missing stream methods
- [ ] `Stream::all`
- [ ] `Stream::any`
- [ ] `Stream::by_ref`
- [ ] `Stream::chain`
- [ ] `Stream::cloned`
- [ ] `Stream::cmp`
- [x] `Stream::collect`
- [ ] `Stream::copied`
- [ ] `Stream::count`
- [ ] `Stream::cycle`
- [ ] `Stream::enumerate`
- [ ] `Stream::eq`
- [ ] `Stream::filter`
- [ ] `Stream::filter_map`
- [ ] `Stream::find`
- [ ] `Stream::find_map`
- [ ] `Stream::flat_map`
- [ ] `Stream::flatten`
- [ ] `Stream::fold`
- [ ] `Stream::for_each`
- [ ] `Stream::fuse`
- [ ] `Stream::ge`
- [ ] `Stream::gt`
- [ ] `Stream::inspect`
- [ ] `Stream::is_sorted`
- [ ] `Stream::is_sorted_by`
- [ ] `Stream::is_sorted_by_key`
- [ ] `Stream::last`
- [ ] `Stream::le`
- [ ] `Stream::lt`
- [ ] `Stream::map`
- [ ] `Stream::max`
- [ ] `Stream::max_by`
- [ ] `Stream::max_by_key`
- [ ] `Stream::min`
- [ ] `Stream::min_by`
- [ ] `Stream::min_by_key`
- [ ] `Stream::ne`
- [ ] `Stream::nth`
- [ ] `Stream::partial_cmp`
- [ ] `Stream::partition`
- [ ] `Stream::peekable`
- [ ] `Stream::position`
- [ ] `Stream::product`
- [ ] `Stream::rev`
- [ ] `Stream::rposition`
- [ ] `Stream::scan`
- [ ] `Stream::size_hint`
- [ ] `Stream::skip`
- [ ] `Stream::skip_while`
- [ ] `Stream::step_by`
- [ ] `Stream::sum`
- [x] `Stream::take`
- [ ] `Stream::take_while`
- [ ] `Stream::try_fold`
- [ ] `Stream::try_for_each`
- [ ] `Stream::unzip`
- [ ] `Stream::zip`

## Missing `IntoIter` impls
- [x] `Vec`
- [ ] `Hashmap`
- [ ] `VecDeque`
- [ ] `HashSet`
- [ ] `BtreeMap`
- [ ] `Option<T>`
- [ ] `BinaryHeap`
- [ ] `Result<T, E>`

## Missing `FromStream` impls
- [ ] `FromStream<char> for String`
- [ ] `FromStream<()> for ()`
- [ ] `FromStream<char> for String`
- [ ] `FromStream<String> for String`
- [ ] `FromStream<&'a char> for String`
- [ ] `FromStream<&'a str> for String`
- [ ] `FromStream<Cow<'a, str> for String`
- [ ] `FromStream<char> for Cow<'a, str>`
- [ ] `FromStream<String> for Cow<'a, str>`
- [ ] `FromStream<&'b str> for Cow<'a, str>`
- [ ] `FromStream<T> for Cow<'a, [T]> where T: Clone`
- [ ] `FromStream<A> for Box<[A]>`
- [ ] `FromStream<A> for VecDeque<A>`
- [ ] `FromStream<Result<A, E>> for Result<V, E> where V: FromStream<A>`
- [ ] `FromStream<Option<A>> for Option<V> where V: FromStream<A>`
- [ ] `FromStream<(K, V)> for BTreeMap<K, V> where K: Ord`
- [ ] `FromStream<(K, V)> for HashMap<K, V, S> where K: Eq + Hash, S: BuildHasher + Default`
- [ ] `FromStream<P> for PathBuf where P: AsRef<Path>`
- [ ] `FromStream<T> for BinaryHeap<T> where T: Ord`
- [ ] `FromStream<T> for BTreeSet<T> where T: Ord`
- [ ] `FromStream<T> for LinkedList<T>`
- [x] `FromStream<T> for Vec<T>`
- [ ] `FromStream<T> for HashSet<T, S> where T: Eq + Hash, S: BuildHasher + Default`

**Comment by vertexclique** [comment]
I've documented some things out here:

https://paper.dropbox.com/doc/async-flow--AjJPPnPNL6eLfTQ3FT3Qew3zAQ-qCeBTk0Z0Nk92kn1ifNbQ

It was open for comments.

**Comment by sepiropht** [comment]
I would like to port map, filter and fold :). I like this function in std. 

I'm never seriously  did any async with rust before. @vertexclique the link you posted is an example we should follow to implement all this functions ?

**Comment by sepiropht** [comment]
I would like to port map, filter and fold :). I like this function in std. 

I'm never seriously  did any async with rust before. @vertexclique the link you posted is an example we should follow to implement all this functions ?

**Comment by shekohex** [comment]
I started with simple methods like `any` and `all` here at #132 

@sepiropht you may take a look at my pr, and the `stream::take()` method too.
simply create your struct `Map` and implement `futures::Stream` for it, by `polling` the inner stream and apply the provided function for every item to the stream.

**Comment by sepiropht** [comment]
Ok thanks @shekohex. I will try that

**Comment by vertexclique** [comment]
@sepiropht from my point of view better to concentrate on `FromStream` and `IntoIter` ones are important. Other than that, I suggest following the list of Yoshua for method impls.

**Comment by taiki-e** [comment]
Note that some features on this list are not useful until asynchronous closure is improved.

**Comment by taiki-e** [comment]
@yoshuawuyts 
btw, the following items are unstable.
https://doc.rust-lang.org/nightly/core/iter/trait.Iterator.html#method.is_sorted

> * [ ]  `Stream::is_sorted`
> * [ ]  `Stream::is_sorted_by`
> * [ ]  `Stream::is_sorted_by_key`

https://doc.rust-lang.org/nightly/core/iter/fn.once_with.html

> * [ ]  `once_with`

**Comment by yoshuawuyts** [comment]
@taiki-e thanks, I've updated the issue!

**Comment by sunjay** [comment]
Do you need someone to add the remaining `FromStream` impls? I had a lot of fun doing #207 and would love to contribute more.

**Comment by stjepang** [comment]
@sunjay Go for it :) And thank you!!

**Comment by yoshuawuyts** [comment]
`IntoStream` impls for `std::collections` types (and comparable) aren't possible in `async-std` because our `Stream` impl builds on top of `futures-rs`'s stream impl. We discovered this in https://github.com/async-rs/async-std/pull/213.

Conversations about possibly adding a `IntoStream` type to `futures-core` are happening in https://github.com/rust-lang-nursery/futures-rs/issues/1879, though we may need to open a dedicated issue for this.

As of now I'm going to go ahead and strike out the missing `IntoStream` impls for the collection types because they aren't possible. Though my hope is that we can move the `IntoStream` trait to be part of `futures-core` eventually.

**Comment by sunjay** [comment]
> Do you need someone to add the remaining `FromStream` impls? I had a lot of fun doing #207 and would love to contribute more.

Small update on this: I started to work on it on the weekend and noticed that I need a few more things to be implemented before finishing it off. In particular, it looks like a lot of the `FromStream` implementations use the `Extend` trait. That exists, so I started to work on adding it for a few cases where I needed it (e.g. for `String` and other collections). 

I could also use a few more combinators on `Stream`/`StreamExt`. In particular: `size_hint`, `for_each`, `cloned`/`copied` were all needed for a few of the impls.

I can work around some of these using `fold`, but it would be nice if the implementations could match `std` as much as possible.

P.S. Can I ask why the `Stream` trait was split into `Stream` and `StreamExt`? Was it just to avoid the way that `Iterator` is a huge trait? I personally found it a little confusing that I had to look at something separate like that. Also hiding it in `prelude` doesn't really help with that. Would love to know your thoughts on this. :)

**Comment by sunjay** [comment]
I opened #265 for the impls that should be easy to merge. I also opened #266 to discuss the changes to the `Extend` trait that I found were necessary for some of my impls. Once these two changes land and we get #264 (thanks @montekki!) I should be able to continue with `FromStream`. :smile:

**Comment by sunjay** [comment]
Still working on this. #266 is ready for review.

By the way, looks like pretty much every single `FromIterator` implementation in the standard library uses the `Extend` trait, so I guess I'm adding that too! Woo code reuse! :smile: 

(Also, the standard library is really inconsistent with its use of `size_hint` to optimize allocations in case anyone is looking for an easy PR to `std`.)

**Comment by sunjay** [comment]
The next three methods that would be really nice to have are: `size_hint`, `cloned`, and `copied`. :smile:

**Comment by sunjay** [comment]
And we're done! :tada: 

You should merge #266 before you merge #271. All the impls for `FromStream` can be checked off once those are merged. :tada:

**Comment by sunjay** [comment]
Woohoo! All of `FromStream` can be checked off now. 🎉🎉🎉

**Comment by yoshuawuyts** [comment]
@sunjay I was thinking of what we could do next for `FromStream`; and perhaps we could polish our  module-level documentation? [`std::iter::FromIterator`](https://doc.rust-lang.org/std/iter/trait.FromIterator.html) seems like it could probably be ported for the most part!

Is that something that you might be interested in doing?

**Comment by sunjay** [comment]
That sounds great! I don't know when I'll next have time to work on this, so someone else is welcome to take my place and help out. 😊

**Comment by yoshuawuyts** [comment]
@sunjay I'll file an issue for this then so someone can pick this up! :blush:

**Comment by starsheriff** [comment]
fyi, I start working on `take_while`

**Comment by starsheriff** [comment]
I will take a look at `peekable` next. Hope there is no blocker.

**Comment by starsheriff** [comment]
Btw. It looks like `fold` and `for_each` are already implemented, but not ticked off in the list above.

**Comment by felipesere** [comment]
I would like to try the min-max-family.

**Comment by ktomsic** [comment]
If they're not claimed yet, I'd like to do `Stream::sum()` and `Stream::product()`.

**Comment by stjepang** [comment]
@ktomsic It's all yours, go ahead :)

**Comment by yjhmelody** [comment]
The Stream::rposition maybe need `DoubleEndStream`?
I see std's signature
```rust
fn rposition<P>(&mut self, predicate: P) -> Option<usize>
where
    P: FnMut(Self::Item) -> bool,
    Self: Sized + ExactSizeIterator + DoubleEndedIterator, 
```
So are the `rev`

**Comment by felipesere** [comment]
I've started to play with the Double ended iterators

**Comment by felipesere** [comment]
@yoshuawuyts I think we can cross of a bunch of things here, unless it only considers `stable` features?
